### ISSUE TITLE
Rewrite in AVR ASM to enable 2MBps serial

### ISSUE THREAD

**Issue by Villadelfia** [FlorentFlament/ym2149-sndh#2]
Since I really wanted to be able to play those 22kHz tunes, I decided to rewrite this in AVR assembly: 
[YM2149.zip](https://github.com/FlorentFlament/ym2149-sndh/files/3234998/YM2149.zip)

It's setup to use a 2MBps baudrate and I've enlarged the buffer to 2044 bytes, leaving 4 bytes for the stack, which is the maximum size it will need in this implementation.

Feel free to add it to the repo if you want.

**Comment by FlorentFlament** [comment]
Hi @Villadelfia, nice one !
Have you been able to play your 22kHz tunes ?
I like both versions, I'll try to add you version of the firmware so that we be able to choose which one we want to use.

**Comment by Villadelfia** [comment]
It seems my YM2149 chip is somewhat defective, it gets garbled if I update the registers too quickly though a logic analyzer reports that everything is working as intended.

I have continued working on it though, and made an even faster version by ditching some features:

1. I massively simplified the RX state machine at the cost of error checking.
    - It begins in the RX_XMIT state, which loads the free size into UDR0 (you can load 2 bytes into UDR0 without a pause because the USART buffer is 2 bytes.)
    - Then it goes into the RX_RCV state, at which point it assumes it will receive as many bytes as there are free in the buffer, at which point it resets to RX_XMIT and puts the current free size in the buffer. No ACK is sent anymore.
    - The RX interrupt takes 19-20 clock cycles, 1250ns at most. Depending on the size of the packets, receiving 22kHz data is certainly within the realm of possibility.
    - This requires that the sending program ignores free_cnt sizes of 0 (just read again), and it requires it to send as many bytes as it can up to the free_cnt amount. Because an arduino resets anyway when you connect, it's no big deal that it will stall at the end of the song.
2. I saved a cycle here and there in the SMP_STATE machine by having registers that contain frequently used data always available.
3. The routine to write out to the YM2149 has been massively improved to the point of always needing NOPs to slow down for the YM2149. I've done this by just writing the address and value to PORTC and PORTD without massaging the data. This is possible because writing to pins that are configured for other peripherals (the timers and usart) is simply ignored.
4. I've optimized the sound level lookup to being done in a mere 2 cycles.
5. The buffer has been increased to 2046 bytes, basically the entire RAM minus room for one interrupt call.

This current version can be found here: 
[YM2149.zip](https://github.com/FlorentFlament/ym2149-sndh/files/3244055/YM2149.zip)

At this point, the main limiter is not the rate at which the buffer is filled, but the rate at which it is consumed. I believe there is still a lot of room for improvement in the SMP_STATE machine, but this would have to be done by making some assumptions about the buffer. I am still experimenting with it. If I manage to speed it up even more I will let you know.

**Comment by Villadelfia** [comment]
Addendum: I've tested the approach and it works, one caveat: You will need to append the following sequence of bytes to the end of the stream: `FFFFFE` (FFFF clock cycles, address FE) 

Otherwise it doesn't know when to stop. With this approach, the code could in theory update a register every 4 clock cycles of the YM2149 so I think just about any tune should now be playable.

At this point I don't think any significant improvements can be made without some ingenious tricks.

Btw, this code is written in ATMEL studio, and can be assembled on unix systems using the avra assembler.

[YM2149.zip](https://github.com/FlorentFlament/ym2149-sndh/files/3244118/YM2149.zip)

**Comment by FlorentFlament** [comment]
Cool,
I've got to rebuild my YM2149 board (I reused the board for other stuff), then I'll play with your code.
I should make a PCB at some point.
Thanks for the code @Villadelfia !

**Comment by Villadelfia** [comment]
I actually have a pcb designed and on the way. If you want I'd gladly send you one of the extras by way of saying thanks for the original code. You can send me an email at randy.thiemann@student.pxl.be

**Comment by FlorentFlament** [comment]
Oh, that would be awesome. I'll send you my details by email.
Thanks

**Comment by FlorentFlament** [comment]
Thanks @Villadelfia, I added your code to the repo (see https://github.com/FlorentFlament/ym2149-sndh/commit/7f73dd901fddac011d7c14c8fbb1fdca869e89a8)

You can build and flash your fast firmware with the following command:
```
$ make flash-fast
```

And you can play an sndh file with:
```
$ sc68 --ym-engine=dump -qqqn <path-to-sndh-file> | python stream-sndh-fast.py
```

I started doing some tests with your fast firmware and my first observations are as follow:
* When playing a 50Hz tune (I tested [Rob Hubbard's Goldrunner][1]), the sound seems less clear (a bit noisy on some notes) than when using the legacy firmware. I tried to lower the UART transmission rate to 1Mbps ; ) , but it didn't change anything. I suspect some timing issues while driving the YM2149 pins.
* I tried to play [Dubmood's Monkey Island][2]. The sound is very noisy, though seemingly a bit better than with the legacy firmware.

What is your outcome when you play these 2 files with you setup ?
* [Goldrunner][1]
* [Monkey Island][2]

[1]: http://sndh.atari.org/sndh/sndh_lf/Hubbard_Rob/Goldrunner.sndh
[2]: http://sndh.atari.org/sndh/sndh_lf/Dubmood/Monkey_Island.sndh

**Comment by Villadelfia** [comment]
I'll let you know as soon as the PCBs arrive, I have all the components except that one... Up to now I just tested using a logic analyzer.

It may be switching too fast for the actual chip, in which case liberal application of NOP instructions in the timing critical section might be useful.

**Comment by Villadelfia** [comment]
Okay, using a breadboard, I've done some testing:

- At 2MBps, the ISR triggers too often, to the point where it hampers the play routine. We can't put the play routine in a CLI/SEI block either because the ATmega328P only has 2 bytes of buffer.
- Lowering this to 1MBps alleviates the problem of play routine hampering mostly, however if we reduce further to 500kBps, we can safely put the play routine in a CLI/SEI block, with the major caveat that this limits what we can play. For some songs, it may be needed to go back to 1MBps and remove the CLI/SEI block.
- At least some of the nops in the play routine can be removed, at least on my chip. I think this is dependent on the silicon lottery. On my chip, I can play as fast as I can toggle the pins.
- I've improved the playback routine significantly, it now updates a register in EIGHT clock cycles. Including a lookup, that comes down to just about 15~ cycles times the registers per update. If the nops are needed for your chip, this will go up by 10 per register.
- Finally, and most importantly, it no uses modulo based counting, instead you must send the relative distance between samples. Turns out that the code actually switches too *fast* for that to work.

In conclusion: It now works perfectly on any song I can throw at it, you just need to change your streaming script to send distance between samples rather than sample time modulo 0xFFFF. I've also included some settings in the header.

[YM2149.zip](https://github.com/FlorentFlament/ym2149-sndh/files/3270153/YM2149.zip)

**Comment by Villadelfia** [comment]
Small edit: Made it so it clears the registers after 0xFE to avoid your ears bleeding.
[YM2149.zip](https://github.com/FlorentFlament/ym2149-sndh/files/3270160/YM2149.zip)

Oh and I hope to get the PCBs in the upcoming week, as soon as I have one built I'll send it in the mail.

**Comment by FlorentFlament** [comment]
Indeed, this sounds much better !
Both [Goldrunner][1] and [Monkey Island][2] sound great !

If I actually assign 0 to RESPECT_SPECS, some notes are missing. It looks like my YM2149F is less tolerant than yours.

I could find some songs that aren't played that great on my chip:
* [Shadow of the Beast][3]
* [Mad Max's digitalized International Karate][4]

With a 1Mbps bitrate, the sound is a bit better, but still super trash .. ; ).

[1]: http://sndh.atari.org/sndh/sndh_lf/Hubbard_Rob/Goldrunner.sndh
[2]: http://sndh.atari.org/sndh/sndh_lf/Dubmood/Monkey_Island.sndh
[3]: http://sndh.atari.org/sndh/sndh_lf/Whittaker_David/Shadow_of_the_Beast.sndh
[4]: http://sndh.atari.org/sndh/sndh_lf/Mad_Max/International_Karate-Digi.sndh

**Comment by Villadelfia** [comment]
Wow, look at the register dump of those songs... Super fast updates on just two registers. I don't think it's possible to get those playing properly using a 'normal' approach.

I do think it's possible to make a firmware *specifically* for those digi songs by constantly iterating over the buffer at a set rate and interpreting the contents as pairs of values for those two registers.

Of course, it would also be possible to do what the raw sndh file is actually doing and generate the register values from much smaller packets by way of a lookup table...

Though at that point you're not really playing songs on the YM2149 anymore, but rather using the chip as a rather poor DAC. I think that most songs that use the YM2149 as 'intended' should work fine now.

**Comment by FlorentFlament** [comment]
Hehe, I think you are right. You improved the player significantly, thanks for that !

Your 3rd point is not clear to me, could you please elaborate ?

Also, I was thinking about emulating the 68000 cpu with the AVR (It looks like the [sndh file format][1] is raw 68000 machine code with some coding conventions). But with its 2KB of RAM, the ATmega328P is a bit short for storing even small sndh files in RAM.

Btw, don't hesitate to do pull requests directly in the repo ; )

[1]: http://sndh.atari.org/fileformat.php

**Comment by Villadelfia** [comment]
What I meant is, the Atari ST also has a very limited memory, those digi-sndh files must certainly be algorithmic in nature. On the other hand, the 68k is quite difficult to emulate on a 328p, for that I think you'd need to step it up an order of magnitude in processing power.

For the corner case of digi files, I think making a raspberry pi hat would be the best plan of action. I might look into that at some point. The circuit would need level shifters but for the rest it's not that much more complex.

**Comment by FlorentFlament** [comment]
Hehe, on the RaspberryPi, we could probably even have the code that drives the YM2149 written in Python ; ) Actually, we may also be able to use a USB-GPIO device, and drive the chip directly from the PC.
